src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs

```diff
@@ -0,0 +1,110 @@
+    <*> mapM convertTypeVarDecl typeVars
+    <*> convertConDecl ident typeVars constrs
+
+-- | Converts all constructors of a data declaration.
+--
+--   We synthesize the return type of the constructor in IR to avoid passing the
+--   identifier and all used type variables to the type converter. This way we
+--   can reuse the existing translation.
+convertConDecl
+  :: IR.DeclIdent     -- ^ The identifier of the data type.
+  -> [IR.TypeVarDecl] -- ^ The type parameters declared by the data type.
+  -> [IR.ConDecl]     -- ^ The constructor declarations of the data type.
+  -> Converter [Agda.Declaration]
+convertConDecl (IR.DeclIdent srcSpan ident) typeVars =
+  mapM $ convertConstructor $ IR.typeApp NoSrcSpan
+                                         (IR.TypeCon srcSpan ident)
+                                         (map IR.typeVarDeclToType typeVars)
+
+-- | Converts a single constructor of a (non-recursive) data type.
+convertConstructor :: IR.Type -> IR.ConDecl -> Converter Agda.Declaration
+convertConstructor retType (IR.ConDecl _ (IR.DeclIdent srcSpan name) argTypes)
+  = Agda.funcSig
+    <$> lookupUnQualAgdaIdentOrFail srcSpan IR.ValueScope name
+    <*> convertConstructorType argTypes retType
+
+-- | Converts a single constructor to a smart constructor, which wraps the normal
+--   constructor in the @Free@ monad using @pure@.
+--
+--   We use Agda's @pattern@ declarations for smart constructors to simplify
+--   pattern matching in proofs.
+generateSmartConDecl :: IR.ConDecl -> Converter Agda.Declaration
+generateSmartConDecl (IR.ConDecl _ (IR.DeclIdent srcSpan name) argTypes) = do
+  smartName <- lookupUnQualAgdaSmartIdentOrFail srcSpan name
+  patternDecl smartName (repeat "x" `zip` argTypes) $ \vars -> do
+    normalName <- Agda.IdentP <$> lookupAgdaValIdentOrFail srcSpan name
+    let pureVal = foldl Agda.appP normalName vars
+    return (Agda.IdentP (Agda.qname' Agda.Base.pure) `Agda.appP` pureVal)
+
+-------------------------------------------------------------------------------
+-- specialized syntax                                                        --
+-------------------------------------------------------------------------------
+
+-- | Creates a declaration for a data type, which is parameterized over @Shape@
+--   and @Pos@.
+freeDataDecl
+  :: Agda.Name          -- ^ Name of the data type
+  -> [Agda.Name]        -- ^ Names of the bound type variables
+  -> [Agda.Declaration] -- ^ List of constructor declarations
+  -> Agda.Declaration
+freeDataDecl dataName typeNames =
+  Agda.dataDecl dataName $ freeArgBinder `snoc` Agda.binding typeNames Agda.set
+
+-- | Creates a new pattern declaration binding variables with the given names
+--   and types.
+patternDecl
+  :: Agda.Name
+  -- ^ Name of the declaration.
+  -> [(String, IR.Type)]
+  -- ^ Types and preferred names for the bound variables.
+  -> ([Agda.Pattern] -> Converter Agda.Pattern)
+  -- ^ Continuation for creating the definition using the new variables.
```

Comment:
I guess that you need to use a new line for the comments here due to the 80 chars rule? ; )

Reply:
That's right. This should conform to [haddocks comment style](https://haskell-haddock.readthedocs.io/en/latest/markup.html#constructors-and-record-fields). `-- ^` always refers to the argument before the comment. If only one `-- ^` is found after a declaration it refers to the whole function/constructor instead of the return type/last argument.
Writing the comments over two lines reduced readability and made them blend together.